src/coreclr/vm/jitinterface.cpp

```diff
@@ -6015,6 +6015,64 @@ void* CEEInfo::getRuntimeTypePointer(CORINFO_CLASS_HANDLE clsHnd)
     return pointer;
 }
 
+
+/***********************************************************************/
+bool CEEInfo::isObjectImmutable(void* objPtr)
+{
+    CONTRACTL{
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    bool isImmutable = false;
+
+    JIT_TO_EE_TRANSITION();
+
+    GCX_COOP();
+    Object* obj = (Object*)objPtr;
+    MethodTable* type = obj->GetMethodTable();
+
+    if (type->IsString() || type == g_pRuntimeTypeClass)
```

Comment:
@VSadov What is your opinion about RuntimeType - should RuntimeType reference count as immutable object where we can omit the barrier on assignment or not?

Reply:
I think we need to establish a pattern for how immortal/frozen objects are created and shared.

- frozen objects cannot refer to ordinary objects. There are multiple reasons for this, but basically it could be challenging to have a cheap guarantee that a frozen object can keep a non-immortal object alive.
- because of above a frozen object cannot have an assignable reference fields, because user will assign random things.

With above limitation we have only 2 cases - do we have writeable fields at all or not.

- have writeable nonreference field -> need to do a release when assigning, for consistency. 
No need for GC barriers though, since object is not going anywhere.

- does not have writeable nonreference field. -> we only need to ensure that we share a complete instance. Current approach with such objects shared via a hashtable that takes a lock on both writes and reads is enough. 

If we ever consider a low-lock table as the way to share such objects, we need to make sure that write of the reference into the table is a release.